pkg/solana/chainwriter/lookups.go

```diff
@@ -0,0 +1,360 @@
+package chainwriter
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	ag_binary "github.com/gagliardetto/binary"
+	"github.com/gagliardetto/solana-go"
+	addresslookuptable "github.com/gagliardetto/solana-go/programs/address-lookup-table"
+	"github.com/gagliardetto/solana-go/rpc"
+
+	"github.com/smartcontractkit/chainlink-solana/pkg/solana/client"
+)
+
+type Lookup interface {
+	Resolve(ctx context.Context, args any, derivedTableMap map[string]map[string][]*solana.AccountMeta, reader client.Reader) ([]*solana.AccountMeta, error)
+}
+
+// AccountConstant represents a fixed address, provided in Base58 format, converted into a `solana.PublicKey`.
+type AccountConstant struct {
+	Name       string
+	Address    string
+	IsSigner   bool
+	IsWritable bool
+}
+
+// AccountLookup dynamically derives an account address from args using a specified location path.
+type AccountLookup struct {
+	Name       string
+	Location   string
+	IsSigner   bool
+	IsWritable bool
+}
+
+// PDALookups generates Program Derived Addresses (PDA) by combining a derived public key with one or more seeds.
+type PDALookups struct {
+	Name string
+	// The public key of the PDA to be combined with seeds. If there are multiple PublicKeys
+	// there will be multiple PDAs generated by combining each PublicKey with the seeds.
+	PublicKey Lookup
+	// Seeds to be derived from an additional lookup
+	Seeds      []Lookup
+	IsSigner   bool
+	IsWritable bool
+	// OPTIONAL: On-chain location and type of desired data from PDA (e.g. a sub-account of the data account)
+	InternalField InternalField
+}
+
+type InternalField struct {
+	Type     reflect.Type
+	Location string
+}
+
+// LookupTables represents a list of lookup tables that are used to derive addresses for a program.
+type LookupTables struct {
+	DerivedLookupTables []DerivedLookupTable
+	StaticLookupTables  []string
```

Comment:
Can't you have an actual solana.PublicKey type here? I think it has json interface methods defined, you get more instant validation if you use the actual type

Reply:
Good point, fixed.